drivers/can/can_mcp25xxfd.c

```diff
@@ -0,0 +1,1550 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT microchip_mcp25xxfd
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/can/transceiver.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/sys/byteorder.h>
+
+#define LOG_LEVEL CONFIG_CAN_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mcp25xxfd_can);
+
+#include "can_mcp25xxfd.h"
+
+#define SP_IS_SET(inst) DT_INST_NODE_HAS_PROP(inst, sample_point) ||
+
+/* Macro to exclude the sample point algorithm from compilation if not used
+ * Without the macro, the algorithm would always waste ROM
+ */
+#define USE_SP_ALGO (DT_INST_FOREACH_STATUS_OKAY(SP_IS_SET) 0)
+
+/* Timeout for changing mode */
+#define MODE_CHANGE_TIMEOUT_USEC (CONFIG_CAN_MCP25XXFD_MODE_CHANGE_TIMEOUT_MSEC * 1000U)
+#define MODE_CHANGE_RETRIES	   100
+
+#define PLLRDY_TIMEOUT_USEC 100000
+#define PLLRDY_RETRIES	   100
+
+/* when a single mailbox is used access to is protected by tx_sem */
+#if MCP25XXFD_TX_FIFO_ITEMS > 1
+#define MCP25XXFD_MAILBOX_LOCK(mutex)   k_mutex_lock(mutex, K_FOREVER)
+#define MCP25XXFD_MAILBOX_UNLOCK(mutex) k_mutex_unlock(mutex)
+#define MCP25XXFD_TXFIFO_LOCK(mutex)    k_mutex_lock(mutex, K_FOREVER)
+#define MCP25XXFD_TXFIFO_UNLOCK(mutex)  k_mutex_unlock(mutex)
+#else
+#define MCP25XXFD_MAILBOX_LOCK(mutex)
+#define MCP25XXFD_MAILBOX_UNLOCK(mutex)
+#define MCP25XXFD_TXFIFO_LOCK(mutex)
+#define MCP25XXFD_TXFIFO_UNLOCK(mutex)
+#endif
+
+/* rxfifo/tefifo access is run in a cooperative thread. */
+/* It won't be interrupted unless SMP is enabled */
+#if CONFIG_SMP
+#define MCP25XXFD_FIFO_LOCK(mutex)   k_mutex_lock(mutex, K_FOREVER)
+#define MCP25XXFD_FIFO_UNLOCK(mutex) k_mutex_unlock(mutex)
+#else
+#define MCP25XXFD_FIFO_LOCK(mutex)
+#define MCP25XXFD_FIFO_UNLOCK(mutex)
+#endif
+
+static int mcp25xxfd_reset(const struct device *dev)
+{
+	const struct mcp25xxfd_config *dev_cfg = dev->config;
+	uint16_t cmd = sys_cpu_to_be16(MCP25XXFD_SPI_INSTRUCTION_RESET);
+
+	const struct spi_buf tx_buf = {.buf = &cmd, .len = sizeof(cmd),};
+	const struct spi_buf_set tx = {.buffers = &tx_buf, .count = 1};
+
+	return spi_write_dt(&dev_cfg->bus, &tx);
+}
+
+static void mcp25xxfd_canframe_to_txobj(const struct can_frame *src, struct mcp25xxfd_txobj *dst)
+{
+	memset(dst, 0, offsetof(struct mcp25xxfd_txobj, DATA));
+
+	if (src->flags & CAN_FRAME_IDE) {
+		dst->SID = src->id >> 18;
+		dst->EID = src->id;
+		dst->IDE = 1;
+	} else {
+		dst->SID = src->id;
+	}
+
+	if (src->flags & CAN_FRAME_BRS) {
+		dst->BRS = 1;
+	}
+
+	if (src->flags & CAN_FRAME_RTR) {
+		dst->RTR = 1;
+	}
+	dst->DLC = src->dlc;
+#if defined(CONFIG_CAN_FD_MODE)
+	if (src->flags & CAN_FRAME_FDF) {
+		dst->FDF = 1;
+	}
+#endif
+	memcpy(dst->DATA, src->data, MIN(can_dlc_to_bytes(src->dlc), CAN_MAX_DLEN));
+}
+
+void *mcp25xxfd_read_reg(const struct device *dev, uint16_t addr, int len)
+{
```

Comment:
As I understand this issue only affects the READ_CRC command:
http://ww1.microchip.com/downloads/en/DeviceDoc/MCP2518FD-Silicon-Errata-Data-Sheet-Clarification-80000789A.pdf

But I do think it will be good to add READ_CRC in the future. I'll add a comment in the code.